### PR TITLE
set up echo VMs using istioctl workload commands

### PR DESCRIPTION
**Pull request by stevenctl**
* ~Set node metadata from meshConfig file provided to agent~
* Workload command generates a valid mesh config yaml readable by agent (top level field should be `defaultConfig`)
* Tests are refactored to use istioctl as [documented](https://istio.io/latest/docs/setup/install/virtual-machine/) for 1.8

### CODE REVIEW

File: pkg/envoy/proxy.go [hunk 1]
```diff
@@ -179,6 +179,13 @@ func (e *envoy) Run(config interface{}, epoch int, abort <-chan error) error {
 
 	/* #nosec */
 	cmd := exec.Command(e.Config.BinaryPath, args...)
+
+	if e.Config.ProxyMetadata != nil {
```
Comment (stevenctl) [inline pkg/envoy/proxy.go:183]: Is there a better mechanism that using Env here? 

Possibly putting it in the bootstrap json? 

https://github.com/istio/istio/blob/07782fd8d2ff0225af494b172dc459e5c9ace04a/pkg/bootstrap/instance.go#L35

Comment (howardjohn) [reply pkg/envoy/proxy.go:183]: I am a bit confused what we are trying to do here? `ProxyMetadata` is really about setting env vars for `istio-agent` to access (for feature flags, and anything `ISTIO_META_` is put in the node bootstrap, but it looks like this is setting it for the envoy process, which shouldn't really do anything?

Comment (stevenctl) [reply pkg/envoy/proxy.go:183]: The workload command sets a lot of metadata in this file and I don't see how we'd ever send that along with discovery requests to istiod. Seems like we load cluster.env/sidecar.env by using env vars – but maybe we need to do this earlier and any vars in agent need to also check values form this file?

Or should we not be setting metadata via proxyconfig if we expect it to be used? Do we expect to load this into the env elsewhere?

Comment (howardjohn) [reply pkg/envoy/proxy.go:183]: Don't we read the cluster.env: https://github.com/istio/istio/blob/e8abb2e69381a8e52d06ae720075d033b190bfd7/tools/packaging/common/istio-start.sh#L37

then later run the agent: https://github.com/istio/istio/blob/e8abb2e69381a8e52d06ae720075d033b190bfd7/tools/packaging/common/istio-start.sh#L124

?

Comment (howardjohn) [reply pkg/envoy/proxy.go:183]: Oh but we never read ProxyMetadata, I see

Comment (howardjohn) [reply pkg/envoy/proxy.go:183]: Yeah its pretty broken, this is part of why we are moving Gateways to be injected, they suffer the same issue.

Its tricky to change the code to read ProxyMetadata instead of env since a lot of them are spaghetti global variables that are read immediately at initialization, before we could possibly read the ProxyMetadata file (so os.SetEnv in the agent isn't going to work right)

Comment (stevenctl) [reply pkg/envoy/proxy.go:183]: So for #29125 we need to move pretty much everything into cluster.env or somehow read this before agent init.. 

Or read this file and have a bunch of conditionals for all the global vars...

Comment (stevenctl) [reply pkg/envoy/proxy.go]: Alright so I updated this to include the proxyconfig.metadata in node metadata. It might not be possible to set feature flags via proxy config, but it feels like setting metadata (network/autoregistration flag) should be allowed.

Comment (stevenctl) [reply pkg/envoy/proxy.go]: I'm also going to use this PR to convert our VM tests to use `istioctl workload ...`

File: pkg/bootstrap/config.go [hunk 2]
```diff
@@ -449,16 +471,13 @@ func getNodeMetaData(envs []string, plat platform.Environment, nodeIPs []string,
 	meta := &model.BootstrapNodeMetadata{}
 	untypedMeta := map[string]interface{}{}
 
-	extractMetadata(envs, IstioMetaPrefix, func(m map[string]interface{}, key string, val string) {
-		m[key] = val
-	}, untypedMeta)
+	if pc != nil && pc.ProxyMetadata != nil {
```
Comment (stevenctl) [inline pkg/bootstrap/config.go:474]: @howardjohn – Are we OK with allowing the mesh config file to set metadata for Pilot to read but not being able to use it to set feature flags for the agent... or would this be even more confusing?

Comment (howardjohn) [reply pkg/bootstrap/config.go:474]: I think this is ok, but do we have a way to also set the feature flags for VMs?

It should also probably have a comment around this  and maybe the code that reads proxy config since its not obvious

File: pkg/test/framework/components/echo/kube/deployment.go [hunk 3]
```diff
@@ -263,49 +263,39 @@ spec:
...
-          {{- range $idx, $line := (Lines $value.Value) }}
-          sudo sh -c 'echo "  {{ $line }}" >> /etc/istio/config/mesh'
-          {{- end }}
-          {{- end }}
-          {{- end }}
+          # TODO reenable this by editing proxy-config before pushing ConfigMap
+          # # Provide a proxyconfig override
+          # 
+          # {{- range $name, $value := $subset.Annotations }}
+          # {{- if eq $name.Name "proxy.istio.io/config" }}
```
Comment (howardjohn) [inline pkg/test/framework/components/echo/kube/deployment.go:291]: shouldn't this be in the `istioctl` command, not here? or it will be, we just don't use the command here yet?

File: pkg/bootstrap/config.go (hunk 2 above)
Comment (stevenctl) [reply pkg/bootstrap/config.go:474]: `cluster.env`/`sidecar.env` files are really the only way. 

Inn the future, we probably won't have a cli flag for every customization to meta/feature flags in the `workload` command, so I think the VM docs deserve a blurb about "further customizations to the VM" and that can talk about what each of these files is responsible for. 

I'll add comments to the code as well.

Comment (howardjohn) [reply pkg/bootstrap/config.go:474]: We could read the proxy.istio.io/config annotation on workload group though possibly? That would require them to create WG manually rather than via CLI though which is ok?

Comment (stevenctl) [reply pkg/bootstrap/config.go:474]: That works but I don't understand what problem it's solving. AFAICT the two issues we have are

1. `istioctl workload entry` sets a lot of metadata in the proxyConfig file. We could just as easily have the output be `cluster.env`. 
1. `istioctl workload entry` generates a proxy config file and it's unclear what parts of it are actually used.

Comment (howardjohn) [reply pkg/bootstrap/config.go:474]: > but does force users to rely on istioctl as a translator from proxyConfig

Isn't that the point? User manages a WorkloadGroup (which includes proxy config annotation, like pods), then istioctl handles the implementation details of how we configure the agent?

Would we solve this by just writing proxyMetadata into sidecar.env? That effectively mirrors what k8s does in the injection. Then the VM case is not special

Comment (stevenctl) [reply pkg/bootstrap/config.go:474]: To solve the "further customizations" this would work to use as a "base" proxyConfig and any of the generated stuff that the `workload entry` command does would get applied on top of this. 

Anything in defaultConfig.proxyMetadata would also be written to `cluster.env` so that it can be used as feature flags. One question is what takes precedence, ISTIO_META values from proxyConfig or from the environment?

File: pkg/test/framework/components/echo/kube/deployment.go (hunk 3 above)
Comment (stevenctl) [reply pkg/test/framework/components/echo/kube/deployment.go:291]: This is what we previously did to support proxyConfig customizations on VMs through the `echo` api. I think this PR will just keep it in test code. I think to allow having the proxyConfig annotation on a WorkloadGroup is a separate enhancement that can be done in another PR.

Comment (stevenctl) [reply pkg/test/framework/components/echo/kube/deployment.go:291]: I don't think our integration actually use this annotation anywhere.

File: pkg/bootstrap/config.go (hunk 2 above)
Comment (howardjohn) [reply pkg/bootstrap/config.go:474]: If we put `defaultConfig.proxyMetadata`  into `cluster.env` then do we even need this code change here to read directly from proxyMetadata? Then we don't even need to worry about precedence?

**Comment by stevenctl** [comment]
Needs some tests but I think `os.Setenv` should work... if we load proxy config first and utilize `.Get` from the `istio.io/env` package to lazily-resolve the env vars. 

We can probably make it so these aren't package-level vars. 

Still feels hacky but it satisfies the requirements of being able to use proxyConfig to configure features and node metadata. Still need to decide which takes precedence.. the local env of the VM/container is probably more specific.

**Comment by howardjohn** [comment]
> package to lazily-resolve the env vars.

I think this will work but I have two concerns. How do we know we changed all variables we care about to lazy? Are we sure we lazily call them *after* we process the proxyEnv? Are we sure we know all the variables that we do care about? Do we have any way to ensure as new settings are added they will be added properly?

I am fairly worried this will lead to hard to debug issues.

Can you clarify what the issue is with putting ProxyConfig.proxyMetadata into cluster.env is? I think you may have previously but I cannot find the discussion.

File: pilot/cmd/pilot-agent/main.go [hunk 4]
```diff
@@ -166,6 +165,23 @@ var (
 		RunE: func(c *cobra.Command, args []string) error {
 			cmd.PrintFlags(c.Flags())
 
+			// must read proxy config first, we set environment vars here that are used elsewhere
+			proxyConfig, err := constructProxyConfig()
+			if err != nil {
+				return fmt.Errorf("failed to get proxy config: %v", err)
+			}
+			for k, v := range proxyConfig.ProxyMetadata {
```
Comment (costinm) [inline pilot/cmd/pilot-agent/main.go:173]: Does it mean we no longer need to inject all the env variables, just the proxy metadata ?

File: pkg/test/framework/components/echo/kube/deployment.go [hunk 5]
```diff
@@ -263,49 +261,25 @@ spec:
         - bash
         - -c
         - |-
-          # Capture all inbound and outbound traffic
```
Comment (costinm) [inline pkg/test/framework/components/echo/kube/deployment.go:266]: Why did you remove this ? Did we drop support for cluster.env and those settings ? Upgrade ?

File: releasenotes/notes/29130.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: installation
+issue:
+  - 29125
+releaseNotes:
+  - |
+    **Improved** istio-agent will send proxy metadata specified in the mesh config file to istiod.
```
Comment (costinm) [inline releasenotes/notes/29130.yaml:8]: Is this something the user needs to be aware of ? AFAIK this is internal detail, no behavior change.

**Review by costinm — COMMENTED** [review]
Can you include in a comment the actual DiscoveryRequest metadata that is sent to Istiod, before and after ? 

The payload size keeps growing - and multiplied with frequent reconnects, workload counts, etc it adds up - we started with few select fields - and it kept growing. We may need to filter a bit. Also to confirm we are only sending env that start with ISTIO_META - and we don't send them 2x.


In general - I love the change, it's one step closer to dynamic changes for those values when agent would request Istiod to send the current value of the default ProxyConfig, and recompute those fields.

**Comment by stevenctl** [comment]
@costinm The way this is written what we send to istiod doesn't change, we still have to have the ISTIO_META prefix for it to be included in that payload. 

@howardjohn I agree this is a pattern that would be very easy to break leading to difficult to debug issues. Moving things away from the package level helps a bit but still isn't great.

The only issue I see with putting proxy meta into cluster.env is the tight coupling with istioctl for all VMs. It seems very easy to assume proxyConfig.proxyMetadata in the file given to agent will have an effect (seems it was made when building the workload commands). This needs to be documented clearly as being a no-op and that environment vars are the sole mechanism for setting metadata/configuring the agent.

Yet another hacky option is a small yaml-parsing Go program that is triggered before agent but the istio-start script to parse proxy config and then we include proxy metadata in the env before starting agent.

File: pkg/test/framework/components/echo/kube/deployment.go (hunk 5 above)
Comment (stevenctl) [reply pkg/test/framework/components/echo/kube/deployment.go:266]: Rather than generating it "by hand" in the deployments command, we're generating it with istioctl and mounting it as a file. This matches what we have documented for users (besides the mounting details).

**Comment by costinm** [comment]
I want to make sure sidecar.env/cluster.env behavior is not changed - it
would be an upgrade risk.

Why do you need 'small yaml parsing' ? Proxy config should be included in
sidecar.env just like it is included
by the injector, so the code in agent should not care if it's running on a
VM or pod.

**Comment by costinm** [comment]
I don't remember of we discussed this in the WG or in private:  one of the
benefits of the XDS proxy in
agent is that it allows us to send default ProxyConfig when it changes as
well as 'on startup'.

The implication is that many of the settings will be dynamic and no longer
require restart of the workload.
The second implication is that we could add ProxyConfig to Sidecar or some
other CRD, so users can
dynamically change other properties and do the customization without
touching the deployment and pod,
which in turn is important for workloads created via K8S API in libraries -
where user can't annotate.

Avoiding restart of pods and VM 'systemclt restart' is a pretty important
improvement IMO, and sending
the ProxyConfig on startup instead of injecting will simplify both VM and
pods - that's the part I like,
removing the .Get() is a step in this direction.

**Comment by costinm** [comment]
BTW - for historic context, the original reason for sending metadata was to
pass environment info that
was not available in pilot. Pilot is watching the pods - so it actually
knows all the env including the
proxy config.  However in the old times the watchers sometimes got the info
after the sidecar
connected, resulting in subtle bugs. It was a quick fix to send everything
in metadata, the XDS code was a
 mess - now it's much cleaner and we can just wait for k8s ( if we don't
already ).

I think long term, to avoid restarts and for other reasons, we will want
Istiod to send the ProxyConfig
to the pod. It would make little sense for the pod to send ProxyConfig meta
back.

**Comment by stevenctl** [comment]
90% of this PR is refactoring tests to use the approach from [the docs](https://istio.io/latest/docs/setup/install/virtual-machine/). 

For now I'll remove the behavior changes to our actual code and follow up with a PR that changes our proxyConfig behavior.

**Comment by stevenctl** [comment]
@nmittler @howardjohn ptal – this is just changing test framework code now. I will follow up with something to actually fix the bug now that we have tests.

File: istioctl/cmd/workload.go [hunk 7]
```diff
@@ -390,11 +391,25 @@ func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.Workloa
...
 	}
 
-	proxyYAML, err := gogoprotomarshal.ToYAML(meshConfig.DefaultConfig)
+	proxyConfig, err := gogoprotomarshal.ToJSONMap(meshConfig.DefaultConfig)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(filepath.Join(dir, "mesh.yaml"), []byte(proxyYAML), filePerms)
+
+	proxyYAML, err := yaml.Marshal(map[string]interface{}{"defaultConfig": proxyConfig})
```
Comment (howardjohn) [inline istioctl/cmd/workload.go:399]: Is this going to be a backwards incompatible change?

Comment (howardjohn) [reply istioctl/cmd/workload.go:399]: oh nevermind, its already broken today?

**Review by howardjohn — APPROVED** [review]

File: istioctl/cmd/workload.go (hunk 7 above)
Comment (stevenctl) [reply istioctl/cmd/workload.go:399]: Yeah, this was broken to begin with.

Comment (stevenctl) [reply istioctl/cmd/workload.go:399]: https://github.com/istio/istio/issues/29125 mentions that the generated mesh config isn't properly formatted.

File: pkg/test/framework/components/echo/kube/instance.go [hunk 8]
```diff
@@ -161,23 +146,215 @@ func newInstance(ctx resource.Context, originalCfg echo.Config) (out *instance,
 	return c, nil
 }
 
-func setupVM(ctx resource.Context, c *instance, cfg echo.Config) error {
+// createVMConfig sets up a Service account,
+func createVMConfig(ctx resource.Context, c *instance, cfg echo.Config) error {
```
Comment (nmittler) [inline pkg/test/framework/components/echo/kube/instance.go:150]: @stevenctl just a nit: there's a lot of code here, would it be worthwhile separating this out into another package and then breaking this down into multiple helper methods?

Comment (stevenctl) [reply pkg/test/framework/components/echo/kube/instance.go:150]: Probably yeah. I'm hoping a good bit of this will be moved into application code in a follow up.

File: pkg/test/framework/components/istio/ingress.go [hunk 9]
```diff
@@ -37,8 +37,8 @@ const (
 	defaultIngressIstioLabel  = "ingressgateway"
 	defaultIngressServiceName = "istio-" + defaultIngressIstioLabel
 
-	eastWestIngressIstioLabel  = "eastwestgateway"
-	eastWestIngressServiceName = "istio-" + eastWestIngressIstioLabel
+	EastWestIngressIstioLabel  = "eastwestgateway"
```
Comment (nmittler) [inline pkg/test/framework/components/istio/ingress.go:40]: Is this used outside of the package?

**Review by nmittler — APPROVED** [review]

File: pkg/test/framework/components/istio/ingress.go (hunk 9 above)
Comment (stevenctl) [reply pkg/test/framework/components/istio/ingress.go]: Oh yeah I had exposed it to do something that ended up being reverted. Thanks!
